File: plugins/plugin-github/rust/src/actions/create_issue.rs
#![allow(missing_docs)]

use async_trait::async_trait;
use serde_json::json;

use super::{ActionContext, ActionResult, GitHubAction};
use crate::error::Result;
use crate::types::CreateIssueParams;
use crate::GitHubService;

pub struct CreateIssueAction;

#[async_trait]
impl GitHubAction for CreateIssueAction {
    fn name(&self) -> &str {
        "CREATE_GITHUB_ISSUE"
    }

    fn description(&self) -> &str {
        "Creates a new issue in a GitHub repository. Use this to report bugs, request features, or track tasks."
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "OPEN_ISSUE",
            "NEW_ISSUE",
            "FILE_ISSUE",
            "REPORT_BUG",
            "CREATE_BUG_REPORT",
            "SUBMIT_ISSUE",
        ]
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_lowercase();

        Ok(text.contains("issue")
            || text.contains("bug")
            || text.contains("report")
            || text.contains("ticket"))
    }

    async fn handler(
        &self,
        context: &ActionContext,
        service: &GitHubService,
    ) -> Result<ActionResult> {
        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let title = context
            .state
            .get("title")
            .and_then(|t| t.as_str())
            .map(|s| s.to_string())
            .unwrap_or_else(|| text.chars().take(100).collect());

        let params = CreateIssueParams {
            owner: context.owner.clone(),
            repo: context.repo.clone(),
            title,
            body: Some(text.to_string()),
            assignees: Vec::new(),
            labels: Vec::new(),
            milestone: None,
        };

        let issue = service.create_issue(params).await?;

        Ok(ActionResult::success(
            format!("Created issue #{}: {}", issue.number, issue.title),
            json!({
                "issue_number": issue.number,
                "html_url": issue.html_url,
            }),
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[tokio::test]
    async fn test_validate_with_issue_keyword() {
        let action = CreateIssueAction;
        let context = ActionContext {
            message: json!({
                "content": { "text": "Create an issue for this bug" }
            }),
            owner: "test".to_string(),
            repo: "test".to_string(),
            state: json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_without_keywords() {
        let action = CreateIssueAction;
        let context = ActionContext {
            message: json!({
                "content": { "text": "Hello world" }
            }),
            owner: "test".to_string(),
            repo: "test".to_string(),
            state: json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }
}

File: plugins/plugin-github/typescript/actions/createIssue.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  type Content,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  type State,
} from "@elizaos/core";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import { GITHUB_SERVICE_NAME, type GitHubService } from "../service";
import { type CreateIssueParams, createIssueSchema, formatZodErrors } from "../types";

const spec = requireActionSpec("CREATE_ISSUE");

const examples: ActionExample[][] = [
  [
    {
      name: "{{user1}}",
      content: {
        text: "Create an issue in my-org/my-repo with title 'Bug: Login fails' and body 'Users cannot log in after update'",
      },
    },
    {
      name: "{{agent}}",
      content: {
        text: "I'll create that issue for you in my-org/my-repo.",
        actions: ["CREATE_GITHUB_ISSUE"],
      },
    },
  ],
  [
    {
      name: "{{user1}}",
      content: {
        text: "Open a new issue titled 'Add dark mode support' with labels 'enhancement' and 'ui'",
      },
    },
    {
      name: "{{agent}}",
      content: {
        text: "Creating a new issue with the title 'Add dark mode support' and the specified labels.",
        actions: ["CREATE_GITHUB_ISSUE"],
      },
    },
  ],
];

export const createIssueAction: Action = {
  name: "CREATE_GITHUB_ISSUE",
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const service = runtime.getService(GITHUB_SERVICE_NAME);
    if (!service) {
      return false;
    }

    const text = (message.content as Content).text?.toLowerCase() ?? "";
    return (
      text.includes("issue") ||
      text.includes("bug") ||
      text.includes("report") ||
      text.includes("ticket")
    );
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> => {
    const service = runtime.getService<GitHubService>(GITHUB_SERVICE_NAME);

    if (!service) {
      logger.error("GitHub service not available");
      if (callback) {
        await callback({
          text: "GitHub service is not available. Please ensure the plugin is properly configured.",
        });
      }
      return { success: false };
    }

    try {
      const content = message.content as Content;
      const text = content.text ?? "";

      const params: CreateIssueParams = {
        owner: (state?.owner as string) ?? service.getConfig().owner ?? "",
        repo: (state?.repo as string) ?? service.getConfig().repo ?? "",
        title: (state?.title as string) ?? text.slice(0, 100),
        body: (state?.body as string) ?? text,
        labels: (state?.labels as string[]) ?? [],
        assignees: (state?.assignees as string[]) ?? [],
      };

      const validation = createIssueSchema.safeParse(params);
      if (!validation.success) {
        const errors = formatZodErrors(validation.error);
        logger.error(`Invalid issue parameters: ${errors}`);
        if (callback) {
          await callback({
            text: `I couldn't create the issue due to missing information: ${errors}`,
          });
        }
        return { success: false };
      }

      const issue = await service.createIssue(params);

      logger.info(`Created issue #${issue.number}: ${issue.title}`);

      if (callback) {
        await callback({
          text: `Created issue #${issue.number}: "${issue.title}"\n\nView it at: ${issue.htmlUrl}`,
        });
      }

      return { success: true };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      logger.error(`Failed to create issue: ${errorMessage}`);

      if (callback) {
        await callback({
          text: `Failed to create the issue: ${errorMessage}`,
        });
      }

      return { success: false };
    }
  },

  examples,
};

export default createIssueAction;
